### ISSUE TITLE
Support using PutObjectAsync return a HttpStatusCode and throw Exception when StatusCode not equal 200

### ISSUE THREAD

**Issue by lengochiep99** [minio/minio-dotnet#1171]
PutObjectAsync return a PutObjectResponse, but property HttpStatusCode and ResponseContent is a internal. I checked log, and HttpStatusCode is 400, but cann't access the attribute ResponseStatusCode

**Comment by ArminShoeibi** [comment]
Yes we need to make this properties public

https://github.com/minio/minio-dotnet/blob/7440578bb43331518123431a6eef908760d23237/Minio/DataModel/Response/GenericResponse.cs#L29

https://github.com/minio/minio-dotnet/blob/7440578bb43331518123431a6eef908760d23237/Minio/DataModel/Response/GenericResponse.cs#L30

**Comment by n-goncalves** [comment]
> Yes, we need to change these properties to public
> 
> https://github.com/minio/minio-dotnet/blob/7440578bb43331518123431a6eef908760d23237/Minio/DataModel/Response/GenericResponse.cs#L29
> 
> https://github.com/minio/minio-dotnet/blob/7440578bb43331518123431a6eef908760d23237/Minio/DataModel/Response/GenericResponse.cs#L30

It's a bit more than that. 

Currently the documentation and the samples that are available here or in https://min.io/docs/minio/linux/developers/dotnet/quickstart.html makes one assume that an upload is successful if no exception is triggered.
This is wrong and dangerous since that assumption can make you lose files as I already wrote in #1176. 
As the person who opened this issue, we've also seen PutObjectResponse with a non 200 HttpStatusCode, 400 or 500 to be precise. 

Example of PutObject documentation taken from README:
```
try 
{
...
// Upload a file to bucket.
 var putObjectArgs = new PutObjectArgs()
  .WithBucket(bucketName)
  .WithObject(objectName)
 .WithFileName(filePath)
 .WithContentType(contentType);
await minio.PutObjectAsync(putObjectArgs).ConfigureAwait(false);
Console.WriteLine("Successfully uploaded " + objectName );
} 
...
```

**Comment by ArminShoeibi** [comment]
> > Yes, we need to change these properties to public
> > https://github.com/minio/minio-dotnet/blob/7440578bb43331518123431a6eef908760d23237/Minio/DataModel/Response/GenericResponse.cs#L29
> > 
> > https://github.com/minio/minio-dotnet/blob/7440578bb43331518123431a6eef908760d23237/Minio/DataModel/Response/GenericResponse.cs#L30
> 
> It's a bit more than that.
> 
> Currently the documentation and the samples that are available here or in https://min.io/docs/minio/linux/developers/dotnet/quickstart.html makes one assume that an upload is successful if no exception is triggered. This is wrong and dangerous since that assumption can make you lose files as I already wrote in #1176. As the person who opened this issue, we've also seen PutObjectResponse with a non 200 HttpStatusCode, 400 or 500 to be precise.
> 
> Example of PutObject documentation taken from README:
> 
> ```
> try 
> {
> ...
> // Upload a file to bucket.
>  var putObjectArgs = new PutObjectArgs()
>   .WithBucket(bucketName)
>   .WithObject(objectName)
>  .WithFileName(filePath)
>  .WithContentType(contentType);
> await minio.PutObjectAsync(putObjectArgs).ConfigureAwait(false);
> Console.WriteLine("Successfully uploaded " + objectName );
> } 
> ...
> ```

I'm sorry, I didn't get that, right? Why is it more than making those modifiers public?

This merge request is ready and makes those modifiers public
https://github.com/minio/minio-dotnet/pull/1195/files

**Comment by n-goncalves** [comment]
Publicizing those modifiers isn't the same as triggering an Exception if the status code isn't 200 which is the title of this  issue.

Is it expected that PutObjectAsync doesn't trigger an Exception but still fails to do the upload ? 
This is mentioned nowhere in the MinIO .NET Client API Documentation. The only way to know that is to read the code in depth ... or experience it the hard way and lose a file due to this.

**Comment by ebozduman** [comment]
@n-goncalves 
You are right. Publicizing those modifiers is a different issue and it has been addressed in #1176 and fixed by #1195.
The missing exception issue you mention above is actually a duplicate of a known problem addressed in #1121 on which the work is still on.
This issue will be kept open as a duplicate of  #1121 for the time being, so that all other duplicate owners can test and verify the fix in their specific scenario.

**Comment by ramin-n** [comment]
> [@n-goncalves](https://github.com/n-goncalves) You are right. Publicizing those modifiers is a different issue and it has been addressed in [#1176](https://github.com/minio/minio-dotnet/issues/1176) and fixed by [#1195](https://github.com/minio/minio-dotnet/pull/1195). The missing exception issue you mention above is actually a duplicate of a known problem addressed in [#1121](https://github.com/minio/minio-dotnet/issues/1121) on which the work is still on. This issue will be kept open as a duplicate of [#1121](https://github.com/minio/minio-dotnet/issues/1121) for the time being, so that all other duplicate owners can test and verify the fix in their specific scenario.

I’m a bit confused about the purpose of the public property ResponseStatusCode. I was looking for a reason for its existence, and while searching I came across issues like this one where developers requested it to be made public. That made me realize it wasn’t originally exposed to developers until those requests.

My question is: how are we expected to use this property?

If we exclude possible bugs in the library and just consider the normal flow:

When an operation succeeds, the status code will always be 200.

When an error occurs, an exception is thrown and we don’t have access to the response or its properties.

So effectively, aren’t we only ever seeing 200 in successful cases? Am I interpreting this correctly?
